### ISSUE TITLE
Allow overriding hostname for secure_channel in Python client

### ISSUE THREAD

**Issue by bluecmd** [grpc/grpc#5697]
Today it's hard (impossible?) to override hostname for the CN certificate validation without going deep into the gRPC internals so adding a hostname_override to grpc.beta.implementations.secure_channel would be useful.

**Comment by nathanielmanistaatgoogle** [comment]
So what led me to leave hostname overriding out of the production API was the direction ["`This *should* be used for testing only.`" where the relevant channel argument is defined](https://github.com/grpc/grpc/blob/d848db1c06123d23888c705530fc012bcc7e9f35/include/grpc/impl/codegen/grpc_types.h#L147).

Are you in a situation in which you need this capability for non-test production code? If you need this capability in a test, are you aware of the [`not_really_secure_channel` function](https://github.com/grpc/grpc/blob/d848db1c06123d23888c705530fc012bcc7e9f35/src/python/grpcio/tests/unit/beta/test_utilities.py#L36) available to test code?

**Comment by bluecmd** [comment]
This is for production code yes. This is for cases where I want to use local daemons on the same system using CN to identify what function they are serving. For example: Imagine I have a "password" daemon, it would have CN=password and I would connect to "localhost".

If for any reason you need to connect to a server without using DNS (for dependency reasons, alternative discovery services or whatnot) you would also need this feature.

**Comment by nathanielmanistaatgoogle** [comment]
@ctiller: Please assess? Is setting `SERVER_HOST_OVERRIDE` in non-test code a good idea?

**Comment by bluecmd** [comment]
Ping

**Comment by aaronjheng** [comment]
Ping

**Comment by nathanielmanistaatgoogle** [comment]
For those of you waiting for this: this capability is [an underdocumented power-user feature in the GA API](https://github.com/grpc/grpc/blob/77e8b714e510c6a3061d17aab8af769a7b45eed4/src/python/grpcio/grpc/__init__.py#L1200) and I doubt that we'll be adding it to the Beta API.

**Comment by bluecmd** [comment]
That's OK for me. I tried following it down the rabbit hole but got lost in grpc_secure_channel_create. Care to show an example how the initialization of the channel would look like?

**Comment by nathanielmanistaatgoogle** [comment]
@bluecmd: how about [something like](https://github.com/grpc/grpc/blob/77e8b714e510c6a30/src/python/grpcio/tests/unit/beta/test_utilities.py#L52)
```
my_channel = grpc.secure_channel(
    my_target, my_channel_credentials,
    options=(('grpc.ssl_target_name_override', my_server_host_override_string,),))
```
?

**Comment by bluecmd** [comment]
Sounds good! That's simple enough for me.

**Comment by nathanielmanistaatgoogle** [comment]
Closing; this has been implemented since 0.15.
